crypto/evp/evp_enc.c

```diff
@@ -336,7 +360,20 @@ int EVP_EncryptUpdate(EVP_CIPHER_CTX *ctx, unsigned char *out, int *outl,
             return 0;
         }
     }
+
+    /* Past this point inl has to be in bytes. */
     i = ctx->buf_len;
+    /*
+     * Aliasing in and out in this way is counter intuitive and only supported
+     * for legacy applications.
+     * Don't rely on this to stay the same in the next major version.
```

Comment:
I'm still unclear on what is it that one shouldn't rely on...

Reply:
In this case IMHO in and out should not alias at all.
But the doc does not say that either....

Reply:
Well, I'd argue that this note is more contradicting than documentation. Indeed, documentation says that function will return error, while this note can be interpreted as some black magic is happens here, but it will disappear. While what's going on in reality is that checks are effectively hardened, so that documentation becomes so to say more right, and there is no sign of it disappearing in the future. As we are not going to support partially overlapping buffers (which are different from fully overlapping buffers), because we simply can't do that, because lower level subroutines don't leave that kind of guarantee. In fact one could always argue, i.e. before documentation was modified to mention partially overlapping buffers, that this was natural limitation and caller could never count on sensible/predictable outcome with overlapping buffers.

Reply:
Yes, in the end the documentation and the implementation should match.
What is going on in the check down below is an implementation detail,
and that should not be relied upon by the caller.